### DISCUSSION TITLE
사용자의 역할 엔티티 개선

### DISCUSSION THREAD

**Discussion by echo724** [i-am-here-too/iht-backend#10] [개선사항 의견]
### 문맥
- 만약 관리자가 출석 관리 대상이 아닌 경우: User != Member
    - 이를 위해 Member에서 역할이 필요 없을 수 있다.(이전에는 구성원이 더 큰 집합이라 사용자는 무조건 구성원이라 생각했기 때문)
    
- 그렇다면 역할을 어떻게 배정할 수 있는가?
    - 관리자 → 그룹의 관리자이기 때문에 그룹에 소속되면 좋을듯 하다.
    - 구성원인 사용자 → 멤버 ID가 필요하다.

### 대안
- GroupUser인 중간 테이블을 생성하는 것은?
    - group_id
    - user_id
    - role
    - related_member_id?: 구성원과 관련된 멤버 ID
    
- 사용자가 구성원(출석체크대상)이 아닐경우: related_member_id = null
    
- 사용자가 그룹의 권한이 있는지?
    - GroupUser의 group_id와 user_id를 통해 권환 확인
    
- 그렇다면 Member는?
    - 사용자에서는 group_id, sub_group_id, name 정도만 가지고 있도록 변경
    
#### 이전(멤버, 유저 부분만 참조)
![image](https://github.com/haru-attendance/haru-backend/assets/50057022/d495ec98-d877-467d-9be6-2dda436e4c24)

#### 이후
![image](https://github.com/haru-attendance/haru-backend/assets/50057022/462e7aa1-4fe0-455b-8923-e8652867fe3c)

**Comment by kong-hana01** [comment]
> 만약 관리자가 출석 관리 대상이 아닌 경우: User != Member

이 부분에 대한 맥락을 잘 이해못했는데, `구성원이지만 유저가 아닌 경우`를 의미하는 건가요?

**Reply by echo724** [reply]
유저인경우 저희가 출석 대상으로 봐서 구성원이라고 생각을 했었는데요, 예를 들어 출석 관리자도 출석 체크의 대상으로 보기 때문에 구성원이어야한다고 했습니다. 하지만 예를어 학원 선생님이 출석 체크를 하려고 할때 선생님은 학생이 아니기 때문에 출석 체크될 필요가 없죠 그래서 사용자 != 구성원인 경우가 있다고 생각했습니다.
그래서 현제 그룹 → 멤버 → 사용자로 관리자를 찾는 것은 불가능 할 수 있죠. (학원 선생님 같이) 따라서 구성원과 사용자를 분리 하고자 중간 테이블을 생각했습니다!

**Comment by echo724** [comment]
![image](https://github.com/haru-attendance/haru-backend/assets/50057022/2e1becd9-19ab-4faa-a88d-c11256db8fb7)
